extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -66,123 +86,174 @@ export class Messaging {
+      app: APP,
+      requestid: requestId,
+      type: MessageType.RESPONSE,
+      name: messageName,
+      data: messageData,
+      rsvp: undefined,
+      error: undefined,
+    });
   }
 
   /**
-   * @param {string} sentinel
-   * @param {string} requestId
-   * @param {string|null} eventType
-   * @param {*} payload
-   * @param {boolean} awaitResponse
+   * @param {number} requestId
+   * @param {string} messageName
+   * @param {string|undefined} reason
    * @private
    */
-  sendMessage_(sentinel, requestId, eventType, payload, awaitResponse) {
-    const message = {
-      sentinel,
-      requestId,
-      type: eventType,
-      payload,
-      rsvp: awaitResponse,
-    };
-    this.target_./*OK*/postMessage(message, this.targetOrigin_);
+  sendResponseError_(requestId, messageName, reason) {
+    this.logError_(
+      TAG + ': sendResponseError_, Message name: ' + messageName, reason);
+    this.sendMessage_({
+      app: APP,
+      requestid: requestId,
+      type: MessageType.RESPONSE,
+      name: messageName,
+      data: null,
+      rsvp: undefined,
+      error: reason,
+    });
   }
 
   /**
-   * @param {string} requestId
-   * @param {*} reason
+   * @param {Message} message
    * @private
    */
-  sendResponseError_(requestId, reason) {
-    this.sendMessage_(
-      responseSentinel_, requestId, 'ERROR', reason, false);
+  sendMessage_(message) {
+    this.target_./*OK*/postMessage(message, this.targetOrigin_);
   }
 
   /**
    * I'm handing an incoming request from Bob. I'll either respond normally
    * (ex: "got it Bob!") or with an error (ex: "I didn't get a word of what
    * you said!").
-   * @param {*} message
+   * @param {Message} message
    * @private
    */
   handleRequest_(message) {
-    dev().info(TAG, 'messaging.js -> handleRequest_');
-    const requestId = message.requestId;
-    const promise = this.requestProcessor_(message.type, message.payload,
-        message.rsvp);
+    dev().info(TAG, 'handleRequest_', message);
+    if (!this.requestProcessor_) {
+      throw new Error(
+        'Cannot handle request because handshake is not yet confirmed!');
+    }
+    const requestId = message.requestid;
+    const msg = dev().assertString(message.name);
     if (message.rsvp) {
+      const promise =
+        this.requestProcessor_(msg, message.data, message.rsvp);
       if (!promise) {
-        this.sendResponseError_(requestId, 'no response');
+        this.sendResponseError_(requestId, msg, 'no response');
         dev().assert(promise,
-          'expected response but none given: ' + message.type);
+          'expected response but none given: ' + message.name);
       }
-      promise.then(payload => {
-        this.sendResponse_(requestId, payload);
+      promise.then(data => {
+        this.sendResponse_(requestId, msg, data);
       }, reason => {
-        this.sendResponseError_(requestId, reason);
+        this.sendResponseError_(requestId, msg, /** @type {string} */(reason));
```

Comment:
Promises normally fail with `Error` object. Converting it to string here is not correct in those cases. Further, the transfer of `Error` object may fail and would definitely be not what we expect.